packages/doenetml-worker-rust/lib-doenetml-core/src/core/state_var_interfaces/text_state_var_interfaces.rs

```diff
@@ -0,0 +1,197 @@
+use crate::{
+    components::prelude::{
+        Dependency, DependencyInstruction, DependencyValueUpdateRequest,
+        RequestDependencyUpdateError, StateVarInterface, StateVarMutableViewTyped,
+        StateVarParameters, StateVarReadOnlyView, StateVarReadOnlyViewTyped,
+    },
+    dependency::DependencySource,
+    ExtendSource,
+};
+
+use super::common::create_dependency_instruction_from_extend_source;
+
+// use super::{
+//     StateVarInterface, StateVarMutableViewTyped, StateVarParameters, StateVarReadOnlyView,
+//     StateVarReadOnlyViewTyped,
+// };
+
+/// A string state variable interface that concatenates all string dependencies.
+///
+/// If `create_dependency_from_extend_source` is true and have an extend source extending this variable,
+/// then prepend the shadowed state variable.
+///
+/// If the state variable has a single dependency that is an essential state variable,
+/// then propagate the `used_default` attribute of the essential state variable.
+#[derive(Debug, Default)]
+pub struct GeneralStringStateVarInterface {
+    string_dependency_values: Vec<StateVarReadOnlyViewTyped<String>>,
+    from_single_essential: bool,
+}
+
+impl StateVarInterface<String> for GeneralStringStateVarInterface {
+    fn return_dependency_instructions(
+        &self,
+        extend_source: Option<&ExtendSource>,
+        parameters: &StateVarParameters,
+    ) -> Vec<DependencyInstruction> {
+        let mut dep_instructs: Vec<DependencyInstruction> = Vec::with_capacity(2);
+
+        if parameters.create_dependency_from_extend_source {
+            if let Some(dep_inst) =
+                create_dependency_instruction_from_extend_source(extend_source, parameters)
```

Comment:
I feel like using `and_then` looks better.

Reply:
Since the only result is a side-effect (appending to a vector), `and_then` feels funny to me. If I try using `and_then` it forces me to return an option. Returning `None` doesn't work, as it cannot infer the type (and there is no type). The best I could do was:
```
.and_then(|dep_inst| {
     dep_instructs.push(dep_inst);
     None::<()>
});
```
But that seemed weird so I reverted it. If I'm thinking about this wrong, let me know.

Reply:
I agree, that doesn't look nice. `.map` would eliminate that, but might be a strange usage...